Documentation/rfc/v3api.proto

```diff
@@ -0,0 +1,276 @@
+    int64 mod_index = 3;
+    // value of the given key
+    bytes value = 4;
+  }
+}
+
+// First all the compare requests are processed.
+// If all the compare succeed, all the success
+// requests will be processed.
+// Or all the failure requests will be processed and
+// all the errors in the comparison will be returned.
+
+// From google paxosdb paper:
+// Our implementation hinges around a powerful primitive which we call MultiOp. All other database
+// operations except for iteration are implemented as a single call to MultiOp. A MultiOp is applied atomically
+// and consists of three components:
+// 1. A list of tests called guard. Each test in guard checks a single entry in the database. It may check
+// for the absence or presence of a value, or compare with a given value. Two different tests in the guard
+// may apply to the same or different entries in the database. All tests in the guard are applied and
+// MultiOp returns the results. If all tests are true, MultiOp executes t op (see item 2 below), otherwise
+// it executes f op (see item 3 below).
+// 2. A list of database operations called t op. Each operation in the list is either an insert, delete, or
+// lookup operation, and applies to a single database entry. Two different operations in the list may apply
+// to the same or different entries in the database. These operations are executed
+// if guard evaluates to
+// true.
+// 3. A list of database operations called f op. Like t op, but executed if guard evaluates to false.
+message TnxRequest {
+  repeated Compare compare = 1;
+  repeated RequestUnion success = 2;
+  repeated RequestUnion failure = 3;
+}
+
+message TnxResponse {
+  optional ResponseHeader header = 1;
+  optional bool succeeded = 2;
+  repeated ResponseUnion responses = 3;
+}
+
+message KeyValue {
+  optional bytes key = 1;
+  // mod_index is the last modified index of the key.
+  optional int64 create_index = 2
+  optional int64 mod_index = 3;
+  // version is the version of the key. A deletion resets
+  // the version to zero and any modification of the key
+  // increases its version.
+  optional int64 version = 4;
+  optional bytes value = 5;
+}
+
+message WatchRangeRequest {
+  // if the range_end is not given, the request returns the key.
+  optional bytes key = 1;
+  // if the range_end is given, it gets the keys in range [key, range_end).
+  optional bytes range_end = 2;
+  // including
+  optional int64 start_index = 3;
+  // excluding
+  optional int64 end_index = 4;
+  // 
+  optional bool progress_notification = 5;
+}
+
+message WatchRangeResponse {
+  optional ResponseHeader header = 1;
+  repeated Event events = 2;
+}
+
+message Evnet {
+  enum EventType {
+    put = 0;
+    delete = 1;
+    expire = 2;
+  }
+  optional EventType event_type = 1;
+  // a put event contains the current key-value
+  // a delete/expire event contains the previous
+  // key-value
+  optional KeyValue kv = 2;
+}
+
+message CompactRequest {
```

Comment:
We're kind of nouning the verbs to wrap the notion of an action. It feels weirdly normal to say "Here's a PUT" (in the sense of the HTTP verb type). 

I guess I could go either way. If COMPACT were an HTTP verb, this would feel right.